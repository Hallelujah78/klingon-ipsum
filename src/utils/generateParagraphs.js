import { generateRandomNumber } from "./utils";

export const generateParagraphs = async (text, setProgressParagraphs) => {
  const textArray = [...text];
  const maxProgressParagraphs = 100;
  const paraLength = [5, 6, 7];

  // turn our array of sentences into a string
  const string = textArray.join("");
  // split our string into an array of sentences and retain the full stops
  const sentenceArray = string.split(/(?<=\.)/);
  // number of sentences in our array
  const numSentences = sentenceArray.length;

  // start of generate an array of paragraph lengths
  const paragraphLengths = [];
  for (let h = 0; h <= numSentences / paraLength[0]; h++) {
    paragraphLengths.push(paraLength[generateRandomNumber(paraLength)]);
  }
  // end of generate an array of sentence lengths
  const paragraphLengthsLength = paragraphLengths.length;
  const batchSizeParagraphs = Math.ceil(paragraphLengthsLength / 100);

  const tempText = [];
  for (let i = 0; i < paragraphLengthsLength; i++) {
    let tempString = "";

    //******** 1 - paragraph length is smaller than or equal to remaining number of sentences in our sentenceArray *******/
    // if length <= sentenceArray.length
    // then just shift lines from sentenceArray to create a paragraph
    if (paragraphLengths[i] <= sentenceArray.length) {
      for (let j = paragraphLengths[i]; j > 0; j--) {
        tempString = tempString + sentenceArray.shift();
      }
      tempText.push(tempString);
    }

    //******** 2 - allow short paragraph lengths where remaining number of sentences is less than length but greater than or equal to 2 *******/
    // if length > sentenceArray.length AND sentenceArray.length >= 2
    // then just create a short paragraph and return false
    if (
      paragraphLengths[i] > sentenceArray.length &&
      sentenceArray.length >= 2
    ) {
      for (let k = sentenceArray.length; k > 0; k--) {
        tempString = tempString + sentenceArray.shift();
      }

      tempText.push(tempString);
    }

    //******** 3 - widowed lines - prevent paragraphs of less than two lines *******/
    // if length > sentenceArray.length
    // and sentenceArray.length <2
    // then put remaining sentences with previous paragraph if it exists
    // return false
    if (
      paragraphLengths[i] > sentenceArray.length &&
      sentenceArray.length < 2
    ) {
      for (let l = sentenceArray.length; l > 0; l--) {
        tempString = tempString + sentenceArray.shift();
      }
      if (tempText.length) {
        tempText[tempText.length - 1] =
          tempText[tempText.length - 1] + tempString;
      } else {
        tempText.push(tempString);
      }
    }

    if (i % batchSizeParagraphs === 0) {
      const newProgress = Math.floor(
        (i / paragraphLengthsLength) * maxProgressParagraphs
      );
      setProgressParagraphs(newProgress);

      await new Promise((resolve) => setTimeout(resolve, 0));
    }

    //********* 0 - number of sentences generated is less than 2 ******/
  }

  setWords("");
  setText("");
  setLengths([]);
  setProgressParagraphs(100);
  setParagraphs(tempText); // text is an array where each element is a number of sentences comprising a paragraph

  setRenderParagraphs(true);
};
