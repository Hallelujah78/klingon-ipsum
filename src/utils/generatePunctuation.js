export const generatePunctuation = async (
  setProgressPunctuation,
  lengths,
  text
) => {
  const tempLengths = lengths; // tempLengths are sentence lengths
  const tempText = text; // an array
  // start of adding punctuation
  const maxProgressPunctuation = 100;
  let totalLength = 0;
  const lengthsLength = tempLengths.length; // number of sentence lengths
  const batchSizePunctuation = Math.floor(lengthsLength / 100); // 1% of the sentence lengths - to update progress

  const textLength = tempText.length - 1;

  // add periods/full stops
  for (let i = 0; i < lengthsLength; i++) {
    if (i % batchSizePunctuation === 0) {
      const newProgress = Math.floor(
        (i / lengthsLength) * maxProgressPunctuation
      );

      setProgressPunctuation(newProgress);
      await new Promise((resolve) => setTimeout(resolve, 0));
    }

    let remainingLength = textLength - totalLength - 1;

    // if remaining length of text array is less than or equal to current sentence length * 2 (allow for spaces), break
    // here we are saying that there is no need to add a period into tempText if remaining length is less than or equal (because we have a final full stop from the initial word array creation step)
    if (remainingLength <= tempLengths[i] * 2) {
      setProgressPunctuation(100);
      return tempText;
    }

    // add a period where there was a space (spaces will always be in odd index positions)
    // with 5000 words, text is an array with 10000 elements, there is a space in between every word

    tempText[totalLength + tempLengths[i] * 2 - 1] = ". "; // index 9 = item 10
    totalLength = totalLength + tempLengths[i] * 2;
  }
  setProgressPunctuation(100);

  return tempText;
};
