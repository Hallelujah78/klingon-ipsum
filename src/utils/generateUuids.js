import { v4 as uuidv4 } from "uuid";

export const generateUuids = (paragraphs) => {
  const paragraphsLength = paragraphs.length;
  const uuids = [];
  for (let i = 0; i < paragraphsLength; i++) {
    uuids.push(uuidv4());
  }
  return uuids;
};
