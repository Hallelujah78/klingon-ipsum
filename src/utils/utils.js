// takes a data structure like an array and generates a random number based on the structure's length - essentially generates a random index
export const generateRandomNumber = (dataStructure) => {
  return Math.floor(Math.random() * dataStructure.length);
};

//set local storage

export const setLocalStorage = (key, value) => {
  localStorage.setItem(key, JSON.stringify(value));
};

//get local storage

export const getLocalStorage = (key) => {
  const value = JSON.parse(localStorage.getItem(key));
  return value ? value : false;
};

// display alert
export const displayAlert = (text, action, ref) => {
  ref.classList.add(`alert-${action}`);
  ref.textContent = text;

  //remove alert
  setTimeout(function () {
    ref.classList.remove(`alert-${action}`);
    ref.textContent = "";
  }, 3000);
};
