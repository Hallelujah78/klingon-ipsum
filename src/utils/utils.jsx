// takes a data structure like an array and generates a random number based on the structure's length - essentially generates a random index
export const generateRandomNumber = (dataStructure) => {
  return Math.floor(Math.random() * dataStructure.length);
};

export const generateText = (wordArray, wordCount) => {
  const sentenceLength = [5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16];
  let text = [];

  // generate the 'text' by adding words and punctuation to an array
  for (let i = 0; i < wordCount; i++) {
    if (i === 0) {
      // first word, don't add space
      text.push(wordArray[generateRandomNumber(wordArray)]);
    }
    if (i === wordCount - 1) {
      text.push(".");
    } else {
      // it's not the first or last word
      text.push(" ");
      text.push(wordArray[generateRandomNumber(wordArray)]);
    }
  }

  //end of generating the text array

  // start of generate an array of sentence lengths
  const lengths = [];
  for (let i = 0; i <= wordCount / sentenceLength[0]; i++) {
    lengths.push(sentenceLength[generateRandomNumber(sentenceLength)]);
  }
  // end of generate an array of sentence lengths

  let totalLength = 0;

  lengths.every((length, index) => {
    let remainingLength = text.length - 1 - totalLength;

    // if remaining length of text array is less than or equal to current sentence length * 2 (allow for spaces), break
    if (remainingLength <= length * 2) {
      return false;
    }

    // add a period where there was a space (spaces will always be in odd index positions)
    text[totalLength + length * 2 - 1] = ". "; // index 9 = item 10
    totalLength = totalLength + length * 2;

    return true;
  });

  return text;
};

//end of generateText

//set local storage

export const setLocalStorage = (key, value) => {
  localStorage.setItem(key, JSON.stringify(value));
};

//get local storage

export const getLocalStorage = (key) => {
  const value = JSON.parse(localStorage.getItem(key));
  return value ? value : false;
};

// start of createParagraphs function

export const createParagraphs = (textArray) => {
  //our random paragraph lengths = number of sentences per paragraph
  const paraLength = [5, 6, 7];

  // turn our array of sentences into a string
  const string = textArray.join("");
  // split our string into an array of sentences and retain the full stops
  const sentenceArray = string.split(/(?<=\.)/);
  // number of sentences in our array
  const numSentences = sentenceArray.length;

  // start of generate an array of paragraph lengths
  const paragraphLengths = [];
  for (let i = 0; i <= numSentences / paraLength[0]; i++) {
    paragraphLengths.push(paraLength[generateRandomNumber(paraLength)]);
  }
  // end of generate an array of sentence lengths

  //
  const text = [];
  console.log(textArray);
  paragraphLengths.every((length) => {
    let tempString = "";

    //******** 1 - paragraph length is smaller than or equal to remaining number of sentences in our sentenceArray *******/
    // if length <= sentenceArray.length
    // then just shift lines from sentenceArray to create a paragraph
    if (length <= sentenceArray.length) {
      for (let i = length; i > 0; i--) {
        tempString = tempString + sentenceArray.shift();
      }
      text.push(tempString);

      return true;
    }

    //******** 2 - allow short paragraph lengths where remaining number of sentences is less than length but greater than or equal to 2 *******/
    // if length > sentenceArray.length AND sentenceArray.length >= 2
    // then just create a short paragraph and return false
    if (length > sentenceArray.length && sentenceArray.length >= 2) {
      for (let i = sentenceArray.length; i > 0; i--) {
        tempString = tempString + sentenceArray.shift();
      }

      text.push(tempString);
      return false;
    }

    //******** 3 - widowed lines - prevent paragraphs of less than two lines *******/
    // if length > sentenceArray.length
    // and sentenceArray.length <2
    // then put remaining sentences with previous paragraph if it exists
    // return false
    if (length > sentenceArray.length && sentenceArray.length < 2) {
      for (let i = sentenceArray.length; i > 0; i--) {
        tempString = tempString + sentenceArray.shift();
      }
      if (text.length) {
        text[text.length - 1] = text[text.length - 1] + tempString;
        return false;
      } else {
        text.push(tempString);
      }
    }
    //********* 0 - number of sentences generated is less than 2 ******/
  });
  return text; // text is an array where each element is a number of sentences comprising a paragraph
};

// display alert
export const displayAlert = (text, action, ref) => {
  ref.classList.add(`alert-${action}`);
  ref.textContent = text;

  //remove alert
  setTimeout(function () {
    ref.classList.remove(`alert-${action}`);
    ref.textContent = "";
  }, 3000);
};
