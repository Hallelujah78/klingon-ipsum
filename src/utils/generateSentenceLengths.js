import { generateRandomNumber } from "./utils";

export const generateSentenceLengths = async (setProgressLengths, words) => {
  const sentenceLength = [5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16];
  const maxProgressLengths = 100;
  const batchSizeLengths = Math.floor(words / sentenceLength[0] / 100);

  // start of generate an array of sentence lengths
  const lengths = [];
  for (let i = 0; i <= words / sentenceLength[0] / 2; i++) {
    lengths.push(sentenceLength[generateRandomNumber(sentenceLength)]);

    if (i % batchSizeLengths === 0) {
      const newProgress = Math.floor(
        (i / (words / sentenceLength[0] / 2)) * maxProgressLengths
      );
      setProgressLengths(newProgress);

      await new Promise((resolve) => setTimeout(resolve, 0));
    }
  }

  setProgressLengths(100);
  return lengths;
};
// end of generate an array of sentence lengths
