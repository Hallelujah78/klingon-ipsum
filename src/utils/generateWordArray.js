import { generateRandomNumber } from "./utils";

export const generateWordArray = async (setProgress, dictionary, words) => {
  const maxProgress = 100;
  const batchSize = Math.floor((words * 2) / 100);

  let text = new Array(words * 2);
  let period = ".";
  let space = " ";

  // generate the 'text' by adding words and punctuation to an array
  for (let i = 0; i < words * 2; i = i + 2) {
    text[i] = dictionary[generateRandomNumber(dictionary)];
    text[i + 1] = space;

    if (i % batchSize === 0) {
      const newProgress = Math.floor((i / (words * 2)) * maxProgress);
      setProgress(newProgress);

      await new Promise((resolve) => setTimeout(resolve, 0));
    }
  }
  //end of generating the text array
  setProgress(100);
  text.pop();
  text.push(period);

  return text;
};
