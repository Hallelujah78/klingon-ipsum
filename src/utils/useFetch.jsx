// react
import { useEffect, useState } from "react";

// utils
import { getLocalStorage, setLocalStorage } from "./utils";

export const useFetch = (url) => {
  const [isLoading, setIsLoading] = useState(true);
  const [dictionary, setDictionary] = useState([]);
  const [isError, setIsError] = useState(false);

  useEffect(() => {
    const getData = async () => {
      let dict = getLocalStorage("klingon");
      if (getLocalStorage("klingon")) {
        setDictionary(getLocalStorage("klingon"));
      } else {
        const response = await fetch(url);
        if (response.status >= 200 && response.status <= 299) {
          dict = await response.json();
          setDictionary(dict.klingon);
          setLocalStorage("klingon", dict.klingon);
        } else {
          setIsError(true);
        }
      }
      setIsLoading(false);
    };
    getData();
  }, [url]);

  return { isLoading, dictionary, isError };
};
