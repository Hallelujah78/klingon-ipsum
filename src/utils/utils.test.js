import { generateRandomNumber } from "./utils";

describe("generateRandomNumber", () => {
  test("test", () => {
    expect([0, 1, 2, 3]).toContain(generateRandomNumber(["a", "b", "c", "d"]));
  });
});
