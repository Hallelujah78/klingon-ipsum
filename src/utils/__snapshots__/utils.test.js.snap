// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateRandomNumber test 1`] = `0`;
